Accountability feature (#81)
autonity/docs.autonity.org: content/docs/reference/api/aut/_index.md

## bond

> Delegates an amount of Newton stake token to a designated validator.
> On successful processing of the method call:
> 
> - the bonded Newton amount is locked in the `msg.Sender`'s Newton account
> - a bonding object for the necessary voting power change is created and tracked in memory until applied at epoch end. At that block point, if the `msg.Sender` is a stake delegator account then Liquid Newton will be minted to the delegator for the bonded stake amount

Reviewer: Note the second list item contains two sentences. So the entire list needs to be comprised of sentences:

```suggestion
Delegates an amount of Newton stake token to a designated validator.
On successful processing of the method call:

- The bonded Newton amount is locked in the `msg.Sender`'s Newton account.
- A bonding object for the necessary voting power change is created and tracked in memory until applied at epoch end. At that block point, if the `msg.Sender` is a stake delegator account then Liquid Newton will be minted to the delegator for the bonded stake amount.
```

## bond

Delegates an amount of Newton stake token to a designated validator.
On successful processing of the method call:

- the bonded Newton amount is locked in the `msg.Sender`'s Newton account

> - a bonding object for the necessary voting power change is created and tracked in memory until applied at epoch end. At that block point, if the `msg.Sender` is a stake delegator account then Liquid Newton will be minted to the delegator for the bonded stake amount.

Reviewer: ```suggestion
- a bonding object for the necessary voting power change is created and tracked in memory until applied at epoch end. At that block point, if the `msg.Sender` is a stake delegator account then Liquid Newton will be minted to the delegator for the bonded stake amount
```

'0x5fe87ee4f61da6e640aec02ce818cdcd30b8cb13'
{{< /tab >}}
-->

> ##  getUnbondingReq

Reviewer 2: This is depreciated, the new way of tracking bonding and unbonding requests is through events

Reviewer 2: Oh actually I think it's because the improved unbonding hasn't been _fully_ documented yet

Author: Yes. I am making edits for that in the `improved-unbonding` branch and PR!

i.e. for each feature addition or update, a discrete PR.

Reviewer 2: got it

| `enode` | `string` | the enode url of the validator node |
| `commissionRate` | `uint256` | the percentage commission that the validator will charge on staking rewards from delegated stake |
| `bondedStake` | `uint256` | the total amount of delegated and self-bonded stake that has been bonded to the validator |
| `selfBondedStake` | `uint256` | the total amount of 'self-bonded' stake that has been bonded to the validator by the validator operator |
| `liquidContract` | `Liquid` | the address of the validator's Liquid Newton contract |
| `liquidSupply` | `uint256` | the total amount of Liquid Newton in circulation |
| `registrationBlock` | `uint256` | the block number in which the registration of the validator was committed to state|
| `totalSlashed` | `uint256` | the total amount of stake that a validator has had slashed for accountability and omission faults since registration |

> | `jailReleaseBlock` | `uint256` | the block number at which a validator jail period applied for an accountability or omission fault ends. (The validator can be re-activated after this block height.) |

Reviewer: ```suggestion
| `jailReleaseBlock` | `uint256` | the block number at which a validator jail period applied for an accountability or omission fault ends (the validator can be re-activated after this block height) |
```

| `provableFaultCount` | `uint256` | a counter of the number of times that a validator has been penalised for accountability and omission faults since registration |

> | `ValidatorState` | `state` | the state of the validator. `ValidatorState` is an enumerated type with enumerations: `active`, `paused`, `jailed` |

Reviewer: ```suggestion
| `ValidatorState` | `state` | the state of the validator, an enumerated type with values: `active`, `paused`, `jailed` |
```

Unbonds an amount of Liquid Newton stake token from a designated validator.

On successful processing of the method call:

- the designated amount of Liquid Newton amount is burnt if the stake being unbonded is _delegated_ and *not* _self-bonded_ stake, 
- the unbonding period begins,
- an unbonding object for the necessary voting power change is created and tracked in memory until applied at the end of the epoch in which the unbonding period expires. At that block point Newton redemption occurs and due Newton is minted to the staker's Newton account.

Constraint checks:

- the `validator` address is registered as a validator
- if the `msg.Sender` is a stake delegator account, then Liquid Newton balance and supply checks are applied:
    - the Liquid Newton balance of the account submitting  the `unbond()` method call has a balance `>=` to the `amount` being unbonded
    - if the validator is in the current consensus committee, then the total supply of Liquid Newton remaining after the unbonding will be `> 0` (preventing the case of a committee member having a `0` supply of LNTN).
- if the `msg.Sender` is a validator `treasury` account:

>     - the `selfBondedStake` balance is`>=` to the `amount` being unbonded

Reviewer: Improper use of a list

Author: I have collapsed this list into a paragraph so:

>Constraint checks are applied. The  `validator` address provided is verified as a registered validator address and the requested unbonding amount is checked to verify it is `<=` to the `msg.sender`'s bonded stake amount. For delegated stake this is done by checking the `msg.Sender`'s Liquid Newton balance is `>=` to the requested amount, and for self-bonded stake this is done by checking the validator's `selfBondedStake` balance is`>=` to the requested unbonding amount.

I have left the preceding list beginning "On successful processing of the method call:" because it is expressing a sequence.

Done in commit 3abf4756763bf6722e4298e9b85af71a0856b3d2

Reviewer: Much more readable but capitalization needs to be revised, I shall propose a fix shortly.
